### PR TITLE
Key vault work - RNG and rotation policy

### LINKED ISSUE #3513: Implement /rotationpolicy API

### LINKED ISSUE #3551: Implement /rng API

### PR DESCRIPTION
**Pull request by gearama**
closes #3513
closes #3551 

Implement rotation policy read write and the RNG API. 

# Pull Request Checklist

Please leverage this checklist as a reminder to address commonly occurring feedback when submitting a pull request to make 
Make sure your PR can be reviewed quickly:

See the detailed list in the [contributing guide](https://github.com/Azure/azure-sdk-for-cpp/blob/main/CONTRIBUTING.md#pull-requests).

- [x] [C++ Guidelines](https://azure.github.io/azure-sdk/cpp_introduction.html)
- [x] Doxygen docs
- [x] Unit tests
- [x] No unwanted commits/changes
- [x] Descriptive title/description
  - [x] PR is single purpose
  - [x] Related issue listed
- [x] Comments in source
- [x] No typos
- [x] Update changelog
- [x] Not work-in-progress
- [x] External references or docs updated
- [x] Self review of PR done
- [x] Any breaking changes?

### CODE REVIEW

File: sdk/keyvault/azure-security-keyvault-keys/inc/azure/keyvault/keys/key_client_models.hpp [hunk 1]
```diff
@@ -899,4 +905,107 @@ namespace Azure { namespace Security { namespace KeyVault { namespace Keys {
         Azure::Core::Context const& context = Azure::Core::Context());
   };
 
+  enum LifetimeActionType
```
Comment (vhvb1989) [inline sdk/keyvault/azure-security-keyvault-keys/inc/azure/keyvault/keys/key_client_models.hpp:908]: use `enum class` instead of `enum`

Comment (vhvb1989) [inline sdk/keyvault/azure-security-keyvault-keys/inc/azure/keyvault/keys/key_client_models.hpp:908]: Add docs for this

File: sdk/keyvault/azure-security-keyvault-keys/inc/azure/keyvault/keys/key_client_models.hpp [hunk 2]
```diff
@@ -899,4 +905,107 @@ namespace Azure { namespace Security { namespace KeyVault { namespace Keys {
...
+   * @brief A condition to be satisfied for an action to be executed.
+   */
+  struct LifetimeActionsTrigger final
+  {
+    /**
+     * @brief Time after creation to attempt to rotate. It only applies to rotate. It will be in
+     * ISO 8601 duration format. Example: 90 days : "P90D"
+     *
+     */
+    Azure::Nullable<std::string> TimeAfterCreate;
```
Comment (vhvb1989) [inline sdk/keyvault/azure-security-keyvault-keys/inc/azure/keyvault/keys/key_client_models.hpp:934]: Is this the name used in other languages as well? I wonder it does not mention `rotate` on the name.

File: sdk/keyvault/azure-security-keyvault-keys/inc/azure/keyvault/keys/key_client_options.hpp [hunk 3]
```diff
@@ -401,4 +401,17 @@ namespace Azure { namespace Security { namespace KeyVault { namespace Keys {
...
+   * @brief Get Random Bytes options
+   *
+   */
+  struct GetRandomBytesOptions final
+  {
+    /**
+     * @brief The requested number of random bytes.
+     *
+     */
+    int32_t Count;
```
Comment (vhvb1989) [inline sdk/keyvault/azure-security-keyvault-keys/inc/azure/keyvault/keys/key_client_options.hpp:414]: we use uint64_t for parameters that are sent to the network.  Is there a reason why we need int32_t?

File: sdk/keyvault/azure-security-keyvault-keys/src/key_client.cpp [hunk 4]
```diff
@@ -381,6 +381,59 @@ Azure::Response<KeyVaultKey> KeyClient::ImportKey(
   return Azure::Response<KeyVaultKey>(std::move(value), std::move(rawResponse));
 }
 
+Azure::Response<KeyRotationPolicy> KeyClient::GetKeyRotationPolicy(
+    std::string const& name,
+    Azure::Core::Context const& context) const
+{
+  // Request with no payload
+  auto request = CreateRequest(HttpMethod::Get, {_detail::KeysPath, name, "rotationpolicy"});
```
Comment (vhvb1989) [inline sdk/keyvault/azure-security-keyvault-keys/src/key_client.cpp:389]: move `rotationpolicy` to a constant. It is used in more than one place within this file

File: sdk/keyvault/azure-security-keyvault-keys/inc/azure/keyvault/keys/key_client.hpp [hunk 5]
```diff
@@ -398,11 +403,55 @@ namespace Azure { namespace Security { namespace KeyVault { namespace Keys {
...
+    /**
+     * @brief Get the requested number of bytes containing random values.
+     *
+     * @remark Get the requested number of bytes containing random values from a managed HSM.
+     *
+     * @param options The request object to get random bytes.
+     * @param context A #Azure::Core::Context controlling the request lifetime.
+     * @return Azure::Response<std::vector<uint8_t>>
+     */
+    Azure::Response<std::vector<uint8_t>> GetRandomBytes(
```
Comment (vhvb1989) [inline sdk/keyvault/azure-security-keyvault-keys/inc/azure/keyvault/keys/key_client.hpp:451]: IIRC, we were going to wrap the vector<uint8_t> into a `GetRandomBytesResults` model.  right? @heaths

File: sdk/keyvault/azure-security-keyvault-keys/src/key_get_random_bytes.cpp [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+
+std::vector<uint8_t> _detail::GetRandomBytesSerializer::GetRandomBytesResponseDeserialize(
+    Azure::Core::Http::RawResponse const& rawResponse)
+{
+  auto body = rawResponse.GetBody();
+  auto jsonParser = Azure::Core::Json::_internal::json::parse(body);
+
+  auto value = jsonParser[_detail::ValueParameterValue].get<std::string>();
+  return Azure::Core::_internal::Base64Url::Base64UrlDecode(value);
+}
```
Comment (vhvb1989) [inline sdk/keyvault/azure-security-keyvault-keys/src/key_get_random_bytes.cpp:32]:
```suggestion
}

```

File: sdk/keyvault/azure-security-keyvault-keys/src/key_get_random_bytes.cpp [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+
+  payload[_detail::CountPropertiesValue] = options.Count;
+
+  return payload.dump();
+}
+
+std::vector<uint8_t> _detail::GetRandomBytesSerializer::GetRandomBytesResponseDeserialize(
+    Azure::Core::Http::RawResponse const& rawResponse)
+{
+  auto body = rawResponse.GetBody();
```
Comment (vhvb1989) [inline sdk/keyvault/azure-security-keyvault-keys/src/key_get_random_bytes.cpp:27]: Avoid making a copy from the body by just getting a reference
```suggestion
  auto const& body = rawResponse.GetBody();
```

File: sdk/keyvault/azure-security-keyvault-keys/src/key_rotation_policy.cpp [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+
+using namespace Azure::Security::KeyVault::Keys;
+using namespace Azure::Core::Json::_internal;
+using Azure::Core::_internal::PosixTimeConverter;
+
+KeyRotationPolicy _detail::KeyRotationPolicySerializer::KeyRotationPolicyDeserialize(
+    Azure::Core::Http::RawResponse const& rawResponse)
+{
+  auto body = rawResponse.GetBody();
+  return KeyRotationPolicyDeserialize(body);
```
Comment (vhvb1989) [inline sdk/keyvault/azure-security-keyvault-keys/src/key_rotation_policy.cpp:19]:
```suggestion
  return KeyRotationPolicyDeserialize(rawResponse.GetBody());
```

File: sdk/keyvault/azure-security-keyvault-keys/src/key_rotation_policy.cpp [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+}
+
+KeyRotationPolicy _detail::KeyRotationPolicySerializer::KeyRotationPolicyDeserialize(
+    std::vector<uint8_t> const& body)
+{
+  auto jsonParser = Azure::Core::Json::_internal::json::parse(body);
```
Comment (vhvb1989) [inline sdk/keyvault/azure-security-keyvault-keys/src/key_rotation_policy.cpp:25]: nit.  Use `const` whenever you are not expecting to mutate the value.

```suggestion
  auto const jsonParser = Azure::Core::Json::_internal::json::parse(body);
```

File: sdk/keyvault/azure-security-keyvault-keys/test/ut/key_rotation_policy_test_live.cpp [hunk 10]
```diff
@@ -0,0 +1,164 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+#include "gtest/gtest.h"
+
+#include "../../src/private/key_serializers.hpp"
```
Comment (vhvb1989) [inline sdk/keyvault/azure-security-keyvault-keys/test/ut/key_rotation_policy_test_live.cpp:6]: It is some way better to include the folder that contains the private headers with CMake project than handling the path manually.
See how we include private headers in azure core tests here: https://github.com/Azure/azure-sdk-for-cpp/blob/main/sdk/core/azure-core/test/ut/CMakeLists.txt#L100

Then you can include the headers as `#include "private/key_serializer.hpp" and there will be only one place (CMake) to be updated if the path changes

File: sdk/keyvault/azure-security-keyvault-keys/test/ut/key_rotation_policy_test_live.cpp [hunk 11]
```diff
@@ -0,0 +1,164 @@
...
+      }
+    }
+    EXPECT_TRUE(found);
+  }
+}
+
+TEST_F(KeyVaultKeyClient, DISABLED_GetRandomBytes)
+{ // NEED TO DISABLE TEST FOR THE MOMENT.
+  // DUE TO ISSUE WITH CREATE EC HSM TEST WHICH FAILS WITH ACTUAL HSM BEING SET IN THE ENVIRONMENT
+  // VARIABLE FILED BUG 3563 TO FIX IT
```
Comment (vhvb1989) [inline sdk/keyvault/azure-security-keyvault-keys/test/ut/key_rotation_policy_test_live.cpp:64]: Create an issue to keep track of re-enabling this test

File: sdk/keyvault/azure-security-keyvault-keys/test/ut/key_rotation_policy_test_live.cpp [hunk 12]
```diff
@@ -0,0 +1,164 @@
...
+
+  auto serialized = KeyRotationPolicySerializer::KeyRotationPolicySerialize(policy);
+
+  std::string serializedString
+      = "{\"attributes\":{\"expiryTime\":\"P48M\"},\"lifetimeActions\":[{\"action\":{\"type\":"
+        "\"rotate\"},\"trigger\":{\"timeAfterCreate\":\"P18M\"}},{\"action\":{\"type\":\"notify\"},"
+        "\"trigger\":{\"timeBeforeExpiry\":\"P30D\"}}]}";
+
+  EXPECT_EQ(serialized, serializedString);
+}
```
Comment (vhvb1989) [inline sdk/keyvault/azure-security-keyvault-keys/test/ut/key_rotation_policy_test_live.cpp:164]:
```suggestion
}

```

**Comment by vhvb1989** [comment]
I have reviewed C++ details mostly, adding @heaths as reviewer for Key-Vault-specific changes/features.

@gearama Are you planning to create an APIReview to show the new APIs?

File: sdk/keyvault/azure-security-keyvault-keys/inc/azure/keyvault/keys/key_client_models.hpp (hunk 1 above)
Comment (gearama) [reply sdk/keyvault/azure-security-keyvault-keys/inc/azure/keyvault/keys/key_client_models.hpp:908]: done

Comment (gearama) [reply sdk/keyvault/azure-security-keyvault-keys/inc/azure/keyvault/keys/key_client_models.hpp:908]: done

File: sdk/keyvault/azure-security-keyvault-keys/inc/azure/keyvault/keys/key_client_models.hpp (hunk 2 above)
Comment (gearama) [reply sdk/keyvault/azure-security-keyvault-keys/inc/azure/keyvault/keys/key_client_models.hpp:934]: same names in .Net. These names are from the documentation (swagger)

File: sdk/keyvault/azure-security-keyvault-keys/inc/azure/keyvault/keys/key_client_options.hpp (hunk 3 above)
Comment (gearama) [reply sdk/keyvault/azure-security-keyvault-keys/inc/azure/keyvault/keys/key_client_options.hpp:414]: in the same file at the top there is an int64_t ;)  
.Net also uses System.Int32. The documentation specifies an int32 to be used.

File: sdk/keyvault/azure-security-keyvault-keys/src/key_rotation_policy.cpp (hunk 8 above)
Comment (gearama) [reply sdk/keyvault/azure-security-keyvault-keys/src/key_rotation_policy.cpp:19]: done

File: sdk/keyvault/azure-security-keyvault-keys/src/key_rotation_policy.cpp (hunk 9 above)
Comment (gearama) [reply sdk/keyvault/azure-security-keyvault-keys/src/key_rotation_policy.cpp:25]: done

File: sdk/keyvault/azure-security-keyvault-keys/test/ut/key_rotation_policy_test_live.cpp (hunk 11 above)
Comment (gearama) [reply sdk/keyvault/azure-security-keyvault-keys/test/ut/key_rotation_policy_test_live.cpp:64]: issue 3563

File: sdk/keyvault/azure-security-keyvault-keys/test/ut/key_rotation_policy_test_live.cpp (hunk 12 above)
Comment (gearama) [reply sdk/keyvault/azure-security-keyvault-keys/test/ut/key_rotation_policy_test_live.cpp:164]: done

**Comment by gearama** [comment]
yes, we will need to have an archboard review

File: sdk/keyvault/azure-security-keyvault-keys/test/ut/key_rotation_policy_test_live.cpp (hunk 10 above)
Comment (gearama) [reply sdk/keyvault/azure-security-keyvault-keys/test/ut/key_rotation_policy_test_live.cpp:6]: done

File: sdk/keyvault/azure-security-keyvault-keys/src/key_get_random_bytes.cpp (hunk 6 above)
Comment (gearama) [reply sdk/keyvault/azure-security-keyvault-keys/src/key_get_random_bytes.cpp:32]: done

File: sdk/keyvault/azure-security-keyvault-keys/src/key_get_random_bytes.cpp (hunk 7 above)
Comment (gearama) [reply sdk/keyvault/azure-security-keyvault-keys/src/key_get_random_bytes.cpp:27]: done

File: sdk/keyvault/azure-security-keyvault-keys/inc/azure/keyvault/keys/key_client.hpp (hunk 5 above)
Comment (gearama) [reply sdk/keyvault/azure-security-keyvault-keys/inc/azure/keyvault/keys/key_client.hpp:451]: .Net returns a byte[] , but you are correct , in C++ we wrap the byte vector in a response object

**Review by vhvb1989 — APPROVED** [review]
LGTM. Thank you for

**Pull request merged**